linera-core/src/client/mod.rs

```diff
@@ -2469,60 +2356,131 @@ where
                         })?;
                         return Ok(ClientOutcome::WaitForTimeout(RoundTimeout {
                             timestamp,
-                            current_round: manager.current_round,
+                            current_round: info.manager.current_round,
                             next_block_height: info.next_block_height,
                         }));
                     }
                     Err(error) => return Err(error),
                 }
+            } else {
+                // Otherwise we have to re-propose the highest validated block.
+                certificate.executed_block().clone()
             }
-        }
+        } else {
+            let block = if let Some(proposal) = info
+                .manager
+                .requested_proposed
+                .as_ref()
+                .filter(|proposal| proposal.content.round.is_fast())
+            {
+                // The fast block counts as "locked", too. If there is one, re-propose that.
+                proposal.content.block.clone()
+            } else if let Some(block) = self.state().pending_block() {
+                block.clone() // Otherwise we are free to propose our own pending block.
+            } else {
+                return Ok(ClientOutcome::Committed(None)); // Nothing to do.
+            };
+            self.stage_block_execution(block).await?.0
+        };
 
-        // The block we want to propose is either the highest validated, or our pending one.
-        let Some(block) = manager
-            .highest_validated_block()
-            .cloned()
-            .or_else(|| self.state().pending_block().clone())
+        let Some(round) =
+            Self::round_for_new_proposal(&info.manager, &identity, Some(&executed_block.block))
         else {
-            return Ok(ClientOutcome::Committed(None)); // Nothing to propose.
+            let timeout = info.round_timeout().ok_or_else(|| {
+                ChainClientError::BlockProposalError("Cannot propose in the current round.")
+            })?;
+            return Ok(ClientOutcome::WaitForTimeout(timeout));
         };
 
-        let round = match Self::round_for_new_proposal(&info, &identity, Some(&block))? {
-            Either::Left(round) => round,
-            Either::Right(timeout) => return Ok(ClientOutcome::WaitForTimeout(timeout)),
+        let block = executed_block.block.clone();
+        let hashed_value = if round.is_fast() {
+            HashedCertificateValue::new_confirmed(executed_block)
+        } else {
+            HashedCertificateValue::new_validated(executed_block)
         };
+        // Collect the hashed certificate values required for execution.
+        let committee = self.local_committee().await?;
+        // Create the final block proposal.
+        let blobs = self.read_local_blobs(block.published_blob_ids()).await?;
+        let key_pair = self.key_pair().await?;
+        let proposal = if let Some(cert) = info.manager.requested_locked {
```

Comment:
We checked this at the beginning of the method, and then followed with comparison whether `certificate.round == requested_locked.round` – if not, re-executed. I'm not sure I understand why here we check the `requested_locked` again but what we do with it is different 👀

Reply:
I just don't know how to structure the code in a way that we could unify those:
To create the proposal I need to know the round and the blobs. If I wanted to create them inside of the different cases above, I'd have to duplicate those calls in each branch.